### PR TITLE
Fix IKTool reporting to use size of markerErrors returned from IKSolver

### PR DESCRIPTION
**Pull request by aseth1**
Fixes issue #2052 

### Brief summary of changes
Use the size of the vector of marker errors returned by the `InverseKinematicsSolver` and do not assume it is the number of references in the `MarkersReference`. In fact, the number is not necessarily fixed, since a marker is ignored if its reference value is NaN.

### Testing I've completed
Changes have no impact on existing tests. They still pass.

### Looking for feedback on...
Would like @chrisdembia to apply fix on the case he reported.

### CHANGELOG.md (choose one)
- no need to update because this is an internal bug fix

### CODE REVIEW

File: OpenSim/Tools/InverseKinematicsTool.cpp [hunk 1]
```diff
@@ -329,6 +329,10 @@ bool InverseKinematicsTool::run()
                 int worst = -1;
 
                 ikSolver.computeCurrentSquaredMarkerErrors(squaredMarkerErrors);
+                // if a marker is not present in the model or its reference value 
+                // is NaN it is not included. Therefore, size of the errors can
+                // vary from frame to frame.
```
Comment (tkuchida) [inline OpenSim/Tools/InverseKinematicsTool.cpp]: Perhaps an Issue should be opened about this? If `squaredMarkerErrors` changes in size, then it seems like the caller would have trouble monitoring the error of a particular marker (its index in the vector will be changing). Furthermore, I'm not sure `getMarkerNameForIndex()` below (on L353) will report the correct marker name if a marker with a lower index has dropped out. I would expect `computeCurrentSquaredMarkerErrors()` to always return a vector of the same (max) size; the caller would then step through (as on L336–342) and ignore any `NaN`s.

**Comment by tkuchida** [comment]
> Changes have no impact on existing tests. They still pass.

Are tests missing?

**Comment by aseth1** [comment]
@tkuchida Yes, I think we were missing some fundamental tests for how order and missing data are handled by the  `InverseKinematicsSolver`. I added a test case to address it at the Solver level. It doesn't directly affect this issue, but now I can definitively say that the number of markers (in the assembly condition) does not change from Frame to frame. In the case of NaN it is not included in the solve but its marker error is reported as 0. @chrisdembia is correct that incorrect number is being used when computing mean error since the number of markers should be decreased by 1 for very NaN marker.

I will fix that next in the Tool. Thanks for the review so far.

**Comment by chrisdembia** [comment]
@aseth1 I pushed a commit with a fix for clang. I then tested this change in my code and the issue is gone.

File: OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp [hunk 2]
```diff
@@ -456,6 +474,114 @@ void testTrackWithUpdateMarkerWeights()
...
+            const auto& markerName = ikSolver.getMarkerNameForIndex(j);
+            cout << " " << markerName << " error = " << markerErrors[j];
+
+            SimTK_ASSERT_ALWAYS( *namesIter++ != "unused",
+                "InverseKinematicsSolver failed to ignore "
+                "unused marker reference (observation).");
+            
+            if (markerName == "m0") {// marker we biased shoulder the error
+                SimTK_ASSERT_ALWAYS(abs(markerErrors[j]-err) <= tol,
+                    "InverseKinematicsSolver mangled marker oder.");
```
Comment (tkuchida) [inline OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp]: oder -> order (also L577)

**Comment by aseth1** [comment]
The issue related to zero marker error being reported when reference is NaN resulting in incorrect error stats was added in #2058.

**Comment by aseth1** [comment]
@tkuchida and @chrisdembia thanks for the review. I decided to add a method to return the number of markers in use by the solver. This number is fixed during IK (repeated calls to track) and I am now using it in the InverseKinematicsTool. Let me know if you are OK with the change. Thanks!

File: OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp [hunk 3]
```diff
@@ -456,6 +474,120 @@ void testTrackWithUpdateMarkerWeights()
...
 
+
+void testNumberOfMarkersMismatch()
+{
+    cout << 
+        "\ntestInverseKinematicsSolver::testNumberOfMarkersMismatch()"
+        << endl;
+
+    std::unique_ptr<Model> pendulum{ constructPendulumWithMarkers() };
+    Coordinate& coord = pendulum->getCoordinateSet()[0];
```
Comment (chrisdembia) [inline OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp]: Should this be `const Coordinate&`?

File: OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp [hunk 4]
```diff
@@ -456,6 +474,120 @@ void testTrackWithUpdateMarkerWeights()
...
+    for (unsigned i = 0; i < markersRef.getNumFrames(); ++i) {
+        state.updTime() = i*dt;
+        ikSolver.track(state);
+
+        //get the marker errors
+        ikSolver.computeCurrentMarkerErrors(markerErrors);
+
+        int nme = markerErrors.size();
+
+        SimTK_ASSERT_ALWAYS(nme = nm,
```
Comment (chrisdembia) [inline OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp]: This is an assignment, not an equality comparison, and will thus return the value of `nm`.

File: OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp [hunk 5]
```diff
@@ -456,6 +474,120 @@ void testTrackWithUpdateMarkerWeights()
...
+        auto namesIter = usedMarkerNames.begin();
+        for (int j = 0; j < nme; ++j) {
+            const auto& markerName = ikSolver.getMarkerNameForIndex(j);
+            cout << " " << markerName << " error = " << markerErrors[j];
+
+            SimTK_ASSERT_ALWAYS( *namesIter++ != "unused",
+                "InverseKinematicsSolver failed to ignore "
+                "unused marker reference (observation).");
+            
+            if (markerName == "m0") {// marker we biased shoulder the error
```
Comment (chrisdembia) [inline OpenSim/Simulation/Test/testInverseKinematicsSolver.cpp]: Could you rephrase "marker we biased shoulder the error"? I don't understand what this means.

**Comment by aseth1** [comment]
> To be clear, this does not fix #2058, correct?

Correct.

**Comment by aseth1** [comment]
@chrisdembia thank you for pointing out those errors. They should be fixed now.

**Comment by aseth1** [comment]
Thanks @chrisdembia!

**Comment by tkuchida** [comment]
> Let me know if you are OK with the change.

:+1: Travis/osx is sitting at "waiting to be queued" (≈ "to be deprecated"); all other tests :white_check_mark:

**Comment by aseth1** [comment]
Thank you @tkuchida
